src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexCache.cs

```diff
@@ -0,0 +1,199 @@
+        {
+            lock (this.lockObj)
+            {
+                // TODO: Have a mapping of last accessed timestamps so items can age out too
+                // TODO: LRU concept - use the dictionary as a lookup for index positions into an array-like data structure
+
+                // Don't bother adding the entry if it exists already.
+                if (this.cachedAddresses.ContainsKey(indexData.Address))
+                    return;
+
+                // Now check if any entries need to be evicted.
+
+                // The size is the base record plus each balance change.
+                // We need to count it this way because an address may have very
+                // few changes and should therefore have less weight than one
+                // with numerous changes.
+                int newItemCount = indexData.BalanceChanges.Count + 1;
+
+                int itemsToEvict = (this.itemCount + newItemCount) - this.MaxItems;
+
+                while (this.itemCount > 0 && itemsToEvict > 0)
+                {
+                    // For now, evict whichever entry appears first, until LRU functionality added.
+                    string addressToEvict = this.cachedAddresses.Keys.FirstOrDefault();
+
+                    if (addressToEvict == null)
+                        break;
+
+                    // If it's been modified it needs to be persisted first, so changes aren't lost.
+                    if (this.dirtyAddresses.Contains(addressToEvict))
+                    {
+                        this.addressIndexerDataCollection.Upsert(this.cachedAddresses[addressToEvict]);
+                        this.dirtyAddresses.Remove(addressToEvict);
+                    }
+
+                    int itemToEvictSize = this.cachedAddresses[addressToEvict].BalanceChanges.Count + 1;
+
+                    this.cachedAddresses.Remove(addressToEvict);
+                    this.itemCount -= itemToEvictSize;
+                    itemsToEvict -= itemToEvictSize;
+                }
+
+                this.cachedAddresses.Add(indexData.Address, indexData);
+                this.itemCount += newItemCount;
+            }
+        }
+
+        /// <summary>
+        /// Retrieves address data, either the cached version if it exists,
+        /// or directly from the underlying database. If it is a previously
+        /// unseen address an empty record will be created and added to the
+        /// cache.
+        /// </summary>
+        /// <param name="address">The address to retrieve data for.</param>
+        public AddressIndexerData GetOrCreateAddress(string address)
+        {
+            lock (this.lockObj)
+            {
+                this.cachedAddresses.TryGetValue(address, out AddressIndexerData indexData);
+
+                if (indexData != null)
+                    return indexData;
+
+                indexData = this.addressIndexerDataCollection.FindById(address);
+
+                if (indexData == null)
```

Comment:
`indexData = this.addressIndexerDataCollection.FindById(address) ?? new AddressIndexerData() { Address = address, BalanceChanges = new List<AddressBalanceChange>() };`

instead of 3 lines